ocdata/core/adslab.py

```diff
@@ -0,0 +1,271 @@
+        adsorbate_c.rotate(angles[2], v="z", center="COM")
+
+        # Translate adsorbate to binding site.
+        com = adsorbate_c.get_center_of_mass()
+        translation_vector = site - com
+        adsorbate_c.translate(translation_vector)
+
+        # Translate the adsorbate by the scaled normal until it is proximate to the surface
+        self._reposition_adsorbate(adsorbate_c, height_adjustment)
+
+        # Combine adsorbate and surface, and set tags correctly
+        structure = surface_c + adsorbate_c
+        tags = [2] * len(adsorbate_c)
+        final_tags = list(surface_c.get_tags()) + tags
+        structure.set_tags(final_tags)
+
+        # Set pbc and cell.
+        structure.cell = surface_c.cell
+        structure.pbc = [True, True, False]
+
+        return (site, structure)
+
+    def place_adsorbate_on_sites(
+        self,
+        sites: list,
+        num_augmentations_per_site: int = 1,
+        height_adjustment: float = 0.1,
+    ):
+        """
+        Place the adsorbate at the given binding sites.
+        """
+        structures = []
+        for site in sites:
+            for _ in range(num_augmentations_per_site):
+                structures.append(self.place_adsorbate_on_site(site, height_adjustment))
+
+        return structures
+
+    def filter_unreasonable_structures(self, structures: list):
+        """
+        Filter out unreasonable adsorbate-surface systems.
+        """
+        filtered_structures = []
+        for i in structures:
+            site, adslab = i
+            if is_adslab_structure_reasonable(adslab):
+                filtered_structures.append(i)
+        return filtered_structures
+
+    def _reposition_adsorbate(
+        self,
+        adsorbate_atoms: ase.atoms.Atoms,
+        height_adjustment: float = 0.1,
+    ):
+        """
+        Translate the adsorbate along the surface normal so that it is proximate
+        to the surface but there is no atomic overlap by iteratively moving the
+        adsorbate away from the surface, along its normal, until there is no
+        more overlap.
+
+        Args:
+            adsorbate_atoms (ase.atoms.Atoms): the adsorbate atoms copy which
+                is being manipulated during placement
+            height_adjustment (float): [Agstroms] the added distance at each
+                translation of the adsorbate away from the surface
+        """
+        surface_normal = self.surface.atoms.cell[2].copy()
+        scaled_surface_normal = (
+            surface_normal * height_adjustment / np.linalg.norm(surface_normal)
+        )
+
+        ## See which atoms are closest
+        surface_atoms = self.surface.atoms[
+            [idx for idx, tag in enumerate(self.surface.atoms.get_tags()) if tag == 1]
+        ]
+        surface_atoms_tiled = custom_tile_atoms(surface_atoms)
+
+        total_distance_traversed = 0
+        overlap_exists = there_is_overlap(adsorbate_atoms, surface_atoms_tiled)
+        while overlap_exists:
```

Comment:
Tested this functionality and looks good. We might want to think of an alternative way to do this rather than iteratively translating and checking the overlap.